src/hotspot/share/utilities/exceptions.cpp

```diff
@@ -608,4 +611,34 @@ void Exceptions::log_exception(Handle exception, const char* message) {
                          MAX_LEN, exception->print_value_string(),
                          MAX_LEN, message);
   }
+  maybe_log_call_stack(exception, !is_throw_bytecode);
+}
+
+// We don't want to use an OopHandle, or else we may prevent this object from being collected.
+// Whenever a GC happens, this will be cleared by Exceptions::clear_logging_cache().
+static oop _last_logged_exception;
```

Comment:
oh gosh I don't like this at all.  Save the exception string if anything.

Reply:
I've got rid of the caching completely. Now the stack trace is printed only at sites that are actually doing a "throw".

We still have some extraneous stack traces that are printed at the exit of a `finally` block. So if you have code that looks like this:

```
try (A a = new A()) {
  try (B b = new B()) {
    try (C c = new C()) {
        ((Object)(null)).toString();
    }
  }
}
```

You will see the stacktrace 4 times. But I think that's acceptable for now. I added a suggestion for fixing that in comments.